src/Common/FiberStack.h

```diff
@@ -41,49 +40,36 @@ class FiberStack
 
     explicit FiberStack(size_t stack_size_ = default_stack_size) : stack_size(stack_size_)
     {
-        page_size = getPageSize();
+        num_pages = 1 + (stack_size - 1) / page_size;
+        num_bytes = (num_pages + 1) * page_size; /// Add one page at bottom that will be used as guard-page
     }
 
-    boost::context::stack_context allocate() const
-    {
-        size_t num_pages = 1 + (stack_size - 1) / page_size;
-        size_t num_bytes = (num_pages + 1) * page_size; /// Add one page at bottom that will be used as guard-page
+    boost::context::stack_context allocate() const;
 
-        void * vp = ::mmap(nullptr, num_bytes, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
-        if (MAP_FAILED == vp)
-            throw DB::ErrnoException(DB::ErrorCodes::CANNOT_ALLOCATE_MEMORY, "FiberStack: Cannot mmap {}.", ReadableSize(num_bytes));
-
-        /// TODO: make reports on illegal guard page access more clear.
-        /// Currently we will see segfault and almost random stacktrace.
-        if (-1 == ::mprotect(vp, page_size, PROT_NONE))
-        {
-            ::munmap(vp, num_bytes);
-            throw DB::ErrnoException(DB::ErrorCodes::CANNOT_ALLOCATE_MEMORY, "FiberStack: cannot protect guard page");
-        }
+    void deallocate(boost::context::stack_context & sctx) const;
+};
 
-        /// Do not count guard page in memory usage.
-        auto trace = CurrentMemoryTracker::alloc(num_pages * page_size);
-        trace.onAlloc(vp, num_pages * page_size);
+class FixedSizeFiberStackCache;
 
-        boost::context::stack_context sctx;
-        sctx.size = num_bytes;
-        sctx.sp = static_cast< char * >(vp) + sctx.size;
-#if defined(BOOST_USE_VALGRIND)
-        sctx.valgrind_stack_id = VALGRIND_STACK_REGISTER(sctx.sp, vp);
-#endif
-        return sctx;
-    }
+/// Fiber stack allocator with cached pages.
+/// It is used to reduce the number of mmap / munmap / mprotect / sysconf syscalls.
+class FixedSizeFiberStackWithCache
+{
+public:
+    explicit FixedSizeFiberStackWithCache(size_t stack_size_ = FiberStack::default_stack_size);
+    FixedSizeFiberStackWithCache(const FixedSizeFiberStackWithCache &);
+    ~FixedSizeFiberStackWithCache();
 
-    void deallocate(boost::context::stack_context & sctx) const
-    {
-#if defined(BOOST_USE_VALGRIND)
-        VALGRIND_STACK_DEREGISTER(sctx.valgrind_stack_id);
-#endif
-        void * vp = static_cast< char * >(sctx.sp) - sctx.size;
-        ::munmap(vp, sctx.size);
+    boost::context::stack_context allocate();
+    void deallocate(boost::context::stack_context & sctx);
+private:
+    /// Cache for fiber stacks, the stack in the cached is allocated by mmap with guard page.
+    FixedSizeFiberStackCache * cache;
```

Comment:
Why do we need to store pointer on the cache instance? It is singleton.

Reply:
Ok, there are many caches in order to reduce contention.
I think if we do allocate and mprotect not under a lock than we do not have to worry about contention here.